### DISCUSSION TITLE
Depency liveValue cannot be accessed in static library

### DISCUSSION THREAD

**Discussion by thd-bs** [pointfreeco/swift-composable-architecture#1657] [General]
### Description

When using a static library that defines a new `DependencyValues` property and defines its `TestDependencyKey` and `DependencyKey` conformance in two separate extensions, force-accessing the `liveValue` from a test using `DependencyValues.withValue(\.context, .live)` does not work and falls back on the `testValue`.

This only happens with a static library. Doing this with a dynamic framework works as expected.
It also seems to happen just by running the app and using the dependency (which would make sense).

It seems in `_liveValue(_ key: Any.Type)` function of the `OpenExistential` file, that the cast to `DependencyKey` fails.
Probably linked to how static libraries are compiled, not sure you can do anything about it on your side. Maybe if `DependencyKey` did not extend `TestDependencyKey` ? But that's maybe not an option.

I have attached a sample vanilla project that reproduces this issue.

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Accessing the `liveValue` from a static library splitting its `DependencyKey`/`TestDependencyKey` conformance in 2 extension works.

### Actual behavior

Accessing the `liveValue` from a static library splitting its `DependencyKey`/`TestDependencyKey` conformance in 2 extension fallbacks to the `testValue`.

### Steps to reproduce

[DependencyIssue.zip](https://github.com/pointfreeco/swift-composable-architecture/files/9968215/DependencyIssue.zip)

1. Download the attached project.
2. Run the test, it will fail because the `testValue` is accessed instead of the `liveValue`, as expected by the forced live context.
3. Run the app, the `testValue` should also be used.

If you do the same by repackaging the `DoSomethingStatic` library as a dynamic framework, it works as expected.

### The Composable Architecture version information

0.45.0

### Destination operating system

iOS 16.1

### Xcode version information

Xcode 14.1

### Swift Compiler version information

```shell
swift-driver version: 1.62.15 Apple Swift version 5.7.1 (swiftlang-5.7.1.135.3 clang-1400.0.29.51)
Target: arm64-apple-macosx12.0
```

**Comment by mbrandonw** [comment]
Hi @thd-bs, thank you very much for the easily reproducible sample project. I'm seeing the same. I also tried this with the version of `_openExistential` that works in 5.6 with the same effect. 🫤

One thing that does fix it is to move the `DependencyKey` conformance to the same file as the `TestDependencyKey` conformance. But then that's a bummer because it means you couldn't separate the interface from implementation if you want to use static libraries, but at least it does workaround the issue.

This does feel like a Swift/Xcode bug, but it needs more exploration. Maybe a discussion needs to be opened on the Swift forums or an issue on the Swift repo.

**Comment by tgrapperon** [comment]
From a quick test, it seems that the compiler thinks that it doesn't need the file and simply doesn't compile it. Adding something in the same file as `.liveValue` and referencing it in your app's target seems to be enough to make this file build, but I didn't manage to make it works for tests:
```swift
extension DoSomethingClient: DependencyKey {
  public static var token: Void = ()

  public static let liveValue: Self = .init(
    doSomething: { print("Do something") }
  )
}
```
And in your app:
```swift
let _ = DoSomethingClient.token
```
When building a dynamic framework, it doesn't makes assumptions about clients and the file is compiled unconditionally. I don't know if there is a specific attribute to add that could help in this situation.

**Comment by mbrandonw** [comment]
@tgrapperon Wild stuff. So seems certainly an Apple/Swift bug.

**Comment by mbrandonw** [comment]
I filed an [issue](https://github.com/apple/swift/issues/62056) in the Swift repo and feedback on this.

**Comment by mbrandonw** [comment]
This seems to be expected behavior of static libraries. See [here](https://github.com/apple/swift/issues/62056#issuecomment-1311930729). That comment also provides a few potential work arounds.

Since this isn't a bug with the library I'm going to convert it to a discussion.

**Reply by mbrandonw** [reply]
FWIW, adding `-ObjC` to the "other link flags" of the app target does fix it for me.

**Reply by thd-bs** [reply]
Thanks a lot for the follow-up and the workaround.

**Comment by mbrandonw** [comment]
I was told to file a feedback to make Xcode/linker automatically add `-ObjC` when Swift code is detected in a static library. I recommend others duplicate it: FB11788325 ([gist](https://gist.github.com/mbrandonw/8168451acd08bcde6e970422e2f359c2)).

**Comment by acosmicflamingo** [comment]
I created this [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions/1745) before @tgrapperon pointed me to this ticket. This explains quite a bit :)

**Reply by acosmicflamingo** [reply]
@tgrapperon should I add something like this to `Sources/Dependencies/DependencyValues.swift`?
```swift
          runtimeWarn(
            """
            "@Dependency(\\.\(function))" has no live implementation, but was accessed from a \
            live context.

            \(dependencyDescription)

            Every dependency registered with the library must conform to "DependencyKey", and \
            that conformance must be visible to the running application.

            To fix, make sure that "\(typeName(Key.self))" conforms to "DependencyKey" by \
            providing a live implementation of your dependency, and make sure that the \
            conformance is linked with this current application.

            If your dependency is actually providing a live implementation and conformance occurs   <==
            within a static library, using the linker flag '-ObjC' will fix the problem.            <==
            """,
            file: DependencyValues.currentDependency.file ?? file,
            line: DependencyValues.currentDependency.line ?? line
          )
```
